backend-shared/hack/service_account.go

```diff
@@ -0,0 +1,170 @@
+					APIVersion: "rbac.authorization.k8s.io/v1",
+					Kind:       "Role",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: ArgoCDManagerClusterRole,
+				},
+				Rules: ArgoCDManagerNamespacePolicyRules,
+			}
+			_, err := clientset.RbacV1().Roles(namespace).Create(context.Background(), &role, metav1.CreateOptions{})
+			if err != nil {
+				if !apierr.IsAlreadyExists(err) {
+					_, err := clientset.RbacV1().Roles(namespace).Update(context.Background(), &role, metav1.UpdateOptions{})
+					if err != nil {
+						return "", err
+					}
+				}
+			}
+
+			roleBinding := rbacv1.RoleBinding{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "rbac.authorization.k8s.io/v1",
+					Kind:       "RoleBinding",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: ArgoCDManagerClusterRoleBinding,
+				},
+				RoleRef: rbacv1.RoleRef{
+					APIGroup: "rbac.authorization.k8s.io",
+					Kind:     "Role",
+					Name:     ArgoCDManagerClusterRole,
+				},
+				Subjects: []rbacv1.Subject{{
+					Kind:      rbacv1.ServiceAccountKind,
+					Name:      ArgoCDManagerServiceAccount,
+					Namespace: ns,
+				}},
+			}
+			_, err = clientset.RbacV1().RoleBindings(namespace).Create(context.Background(), &roleBinding, metav1.CreateOptions{})
+			if err != nil {
+				if !apierr.IsAlreadyExists(err) {
+					_, err := clientset.RbacV1().RoleBindings(namespace).Update(context.Background(), &roleBinding, metav1.UpdateOptions{})
+					if err != nil {
+						return "", err
+					}
+				}
+			}
+		}
+	}
+	return GetServiceAccountBearerToken(clientset, ns, ArgoCDManagerServiceAccount)
+}
+
+// GetServiceAccountBearerToken will attempt to get the provided service account until it
+// exists, iterate the secrets associated with it looking for one of type
+// kubernetes.io/service-account-token, and return it's token if found.
+func GetServiceAccountBearerToken(clientset kubernetes.Interface, ns string, sa string) (string, error) {
+	var serviceAccount *corev1.ServiceAccount
+	var secret *corev1.Secret
+	var err error
+	err = wait.Poll(500*time.Millisecond, 30*time.Second, func() (bool, error) {
+		serviceAccount, err = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		// Scan all secrets looking for one of the correct type:
+		for _, oRef := range serviceAccount.Secrets {
+			var getErr error
+			secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), oRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, fmt.Errorf("Failed to retrieve secret %q: %v", oRef.Name, getErr)
+			}
+			if secret.Type == corev1.SecretTypeServiceAccountToken {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("Failed to wait for service account secret: %v", err)
+	}
+	token, ok := secret.Data["token"]
+	if !ok {
+		return "", fmt.Errorf("Secret %q for service account %q did not have a token", secret.Name, serviceAccount)
```

Comment:
If you are actually going to create a unit test for checking this errors, it would make your life easier when writing unit-tests, so you can easily do an equality check for a specific error condition. For example:

```go
var secretErr = errors.New("secret err message blah-blah")
```

so you can easily test it:

```go
func someFunc() {
    secretMock.EXPECT().Save(gomock.Eq(&foooBarBar)).Return(nil, secretErr)
 }
```

Of course if you are _not_ going to test this, there is no point going that extra mile. I would just like to make sure you are aware of this, as you use `fmt.Errorf` in a lot's of places in your code.

Reply:
I might be having basic knowledge in this area, so please correct me. I was using `fmt.Errorf` because of its ability to add errors with formatted strings, while when we use an extra method to format those strings `errors.New(Sprintf(""))`  hence less readability, can we not check for the above utility you mentioned? 
Because, both return the same type right(error)?

Reply:
I recommend sticking with `fmt.Errorf` for now, and then if/when we want to do testing of specific errors, to refactor the code at that point.

Reply:
Thank you for that explanation, @drpaneas, noted :taco: